File: src/web-ui/src/infrastructure/api/service-api/LearningProposalAPI.ts
import { api } from './ApiClient';
import { createTauriCommandError } from '../errors/TauriCommandError';

export type LearningProposalSourceKind =
  | 'user_message'
  | 'assistant_text'
  | 'assistant_thinking'
  | 'tool'
  | 'unknown';

export interface LearningProposalSelection {
  selectedText: string;
  turnId: string;
  roundId?: string;
  itemId?: string;
  sourceKind: LearningProposalSourceKind;
}

export interface LearningProposalSource extends LearningProposalSelection {
  sessionId: string;
  workspacePath: string;
  remoteConnectionId?: string;
  remoteSshHost?: string;
}

export type LearningProposalStatus =
  | 'analyzing'
  | 'ready'
  | 'applying'
  | 'applied'
  | 'rejected'
  | 'stale'
  | 'failed';

export type LearningProposalTargetKind = 'memory' | 'skill' | 'agents_md' | 'none';
export type LearningProposalApplyMode = 'memory_note' | 'read_only';

export interface LearningProposalTarget {
  kind: LearningProposalTargetKind;
  applyMode: LearningProposalApplyMode;
  displayName: string;
  identifier?: string;
  filePath?: string;
}

export interface LearningProposalPreview {
  filePath?: string;
  originalContent: string;
  proposedContent: string;
}

export interface LearningProposalError {
  code: string;
  message: string;
}

export interface LearningProposal {
  schemaVersion: number;
  proposalId: string;
  status: LearningProposalStatus;
  source: LearningProposalSource;
  target?: LearningProposalTarget;
  rationale?: string;
  futureUse?: string;
  preview?: LearningProposalPreview;
  baseHash?: string;
  diffHash?: string;
  createdAt: number;
  updatedAt: number;
  error?: LearningProposalError;
}

export interface LearningProposalWorkspaceContext {
  workspacePath?: string;
  remoteConnectionId?: string;
  remoteSshHost?: string;
}

export interface CreateLearningProposalRequest extends LearningProposalWorkspaceContext {
  sessionId: string;
  workspacePath: string;
  source: LearningProposalSelection;
}

export interface GetLearningProposalRequest extends LearningProposalWorkspaceContext {
  proposalId: string;
}

export interface ApproveLearningProposalRequest extends GetLearningProposalRequest {
  baseHash: string;
  diffHash: string;
}

export interface ListLearningProposalsRequest {
  includeResolved?: boolean;
}

export class LearningProposalAPI {
  async list(request: ListLearningProposalsRequest = {}): Promise<LearningProposal[]> {
    try {
      return await api.invoke<LearningProposal[]>('list_learning_proposals', { request });
    } catch (error) {
      throw createTauriCommandError('list_learning_proposals', error, request);
    }
  }

  async create(request: CreateLearningProposalRequest): Promise<LearningProposal> {
    return this.invoke('create_learning_proposal', request);
  }

  async get(request: GetLearningProposalRequest): Promise<LearningProposal> {
    return this.invoke('get_learning_proposal', request);
  }

  async refresh(request: GetLearningProposalRequest): Promise<LearningProposal> {
    return this.invoke('refresh_learning_proposal', request);
  }

  async approve(request: ApproveLearningProposalRequest): Promise<LearningProposal> {
    return this.invoke('approve_learning_proposal', request);
  }

  async reject(request: GetLearningProposalRequest): Promise<LearningProposal> {
    return this.invoke('reject_learning_proposal', request);
  }

  private async invoke<TRequest extends object>(
    command: string,
    request: TRequest,
  ): Promise<LearningProposal> {
    try {
      return await api.invoke<LearningProposal>(command, { request });
    } catch (error) {
      throw createTauriCommandError(command, error, request);
    }
  }
}

export const learningProposalAPI = new LearningProposalAPI();

File: src/apps/desktop/src/api/learning_proposal_api.rs
use bitfun_core::agentic::learning_proposals::{
    get_learning_proposal_service, ApproveLearningProposalRequest, CreateLearningProposalRequest,
    GetLearningProposalRequest, LearningProposal, ListLearningProposalsRequest,
    RefreshLearningProposalRequest, RejectLearningProposalRequest,
};
use log::{error, info};

#[tauri::command]
pub async fn create_learning_proposal(
    request: CreateLearningProposalRequest,
) -> Result<LearningProposal, String> {
    let session_id = request.session_id.clone();
    let turn_id = request.source.turn_id.clone();
    let result = get_learning_proposal_service().create(request).await;
    match result {
        Ok(proposal) => {
            info!(
                "Learning proposal created: proposal_id={}, session_id={}, turn_id={}, status={:?}",
                proposal.proposal_id, session_id, turn_id, proposal.status
            );
            Ok(proposal)
        }
        Err(err) => {
            error!(
                "Failed to create learning proposal: session_id={}, turn_id={}, error={err}",
                session_id, turn_id
            );
            Err(err.to_string())
        }
    }
}

#[tauri::command]
pub async fn get_learning_proposal(
    request: GetLearningProposalRequest,
) -> Result<LearningProposal, String> {
    get_learning_proposal_service()
        .get(&request)
        .await
        .map_err(|err| err.to_string())
}

#[tauri::command]
pub async fn list_learning_proposals(
    request: ListLearningProposalsRequest,
) -> Result<Vec<LearningProposal>, String> {
    get_learning_proposal_service()
        .list(&request)
        .await
        .map_err(|err| err.to_string())
}

#[tauri::command]
pub async fn refresh_learning_proposal(
    request: RefreshLearningProposalRequest,
) -> Result<LearningProposal, String> {
    let proposal_id = request.proposal_id.clone();
    let result = get_learning_proposal_service().refresh(&request).await;
    if let Err(err) = &result {
        error!(
            "Failed to refresh learning proposal: proposal_id={}, error={err}",
            proposal_id
        );
    }
    result.map_err(|err| err.to_string())
}

#[tauri::command]
pub async fn approve_learning_proposal(
    request: ApproveLearningProposalRequest,
) -> Result<LearningProposal, String> {
    let proposal_id = request.proposal_id.clone();
    let result = get_learning_proposal_service().approve(&request).await;
    match result {
        Ok(proposal) => {
            info!(
                "Learning proposal approval handled: proposal_id={}, status={:?}",
                proposal_id, proposal.status
            );
            Ok(proposal)
        }
        Err(err) => {
            error!(
                "Failed to approve learning proposal: proposal_id={}, error={err}",
                proposal_id
            );
            Err(err.to_string())
        }
    }
}

#[tauri::command]
pub async fn reject_learning_proposal(
    request: RejectLearningProposalRequest,
) -> Result<LearningProposal, String> {
    let proposal_id = request.proposal_id.clone();
    let result = get_learning_proposal_service().reject(&request).await;
    if let Err(err) = &result {
        error!(
            "Failed to reject learning proposal: proposal_id={}, error={err}",
            proposal_id
        );
    }
    result.map_err(|err| err.to_string())
}
